SourceX/qol.cpp

```diff
@@ -0,0 +1,377 @@
+/**
+ * @file qol.cpp
+ *
+ * Quality of life features
+ */
+#include "all.h"
+#include "../3rdParty/Storm/Source/storm.h"
+
+DEVILUTION_BEGIN_NAMESPACE
+
+int drawMinX;
+int drawMaxX;
+int highlightItemsMode = 0;
+// 0 = disabled
+// 1 = highlight when alt pressed
+// 2 = hide when alt pressed
+// 3 = always highlight
+bool altPressed = false;
+bool drawXPBar = false;
+bool drawHPBar = false;
+
+int GetTextWidth(char *s)
+{
+	int l = 0;
+	while (*s) {
+		l += fontkern[fontframe[gbFontTransTbl[(BYTE)*s++]]] + 1;
+	}
+	return l;
+}
+
+int GetConfigIntValue(const char *valuename, int base)
+{
+	if (!SRegLoadValue("devilutionx", valuename, 0, &base)) {
+		SRegSaveValue("devilutionx", valuename, 0, base);
+	}
+	return base;
+}
+
+inline void FastDrawHorizLine(int x, int y, int width, BYTE col)
+{
+	memset(&gpBuffer[SCREENXY(x, y)], col, width);
+}
+
+inline void FastDrawVertLine(int x, int y, int height, BYTE col)
+{
+	BYTE *p = &gpBuffer[SCREENXY(x, y)];
+	for (int j = 0; j < height; j++) {
+		*p = col;
+		p += BUFFER_WIDTH;
+	}
+}
+
+inline void FillRect(int x, int y, int width, int height, BYTE col)
+{
+	for (int j = 0; j < height; j++) {
+		FastDrawHorizLine(x, y + j, width, col);
+	}
+}
+
+inline void FillSquare(int x, int y, int size, BYTE col)
+{
+	FillRect(x, y, size, size, col);
+}
+
+void DrawMonsterHealthBar()
+{
+	if (!drawHPBar)
+		return;
+	if (currlevel == 0)
+		return;
+	if (pcursmonst == -1)
+		return;
+
+	int width = 250;
+	int height = 25;
+	int x = 0;          // x offset from the center of the screen
+	int y = 20;         // y position
+	int xOffset = 0;    // empty space between left/right borders and health bar
+	int yOffset = 1;    // empty space between top/bottom borders and health bar
+	int borderSize = 2; // size of the border around health bar
+	BYTE borderColors[] = { 242 /*undead*/, 232 /*demon*/, 182 /*beast*/ };
+	BYTE filledColor = 142;                                             // filled health bar color
+	bool fillCorners = false;                                           // true to fill border corners, false to cut them off
+	int square = 10;                                                    // resistance / immunity / vulnerability square size
+	char *immuText = "IMMU: ", *resText = "RES: ", *vulnText = ":VULN"; // text displayed for immunities / resistances / vulnerabilities

```

Comment:
`const char *` may avoid some "warning: ISO C++ forbids converting a string constant to 'char*'"